ci: Reorganize CI

LecrisUT opened the pull request:
- Rename the CI jobs so that they are more legible in the report
- Fix concurrency on the imported jobs
- Separated pre-commit to its own step
- ~~Add a static-analysis checker. I was experimenting with `sonarcloud` at [`CMake-Template`](https://github.com/LecrisUT/CMake-Template) repo. I think it could be useful to include it as an intermediate to `clang-tidy`. It is also polyglot, so it might pick up on more nieche problems~~. Postponed to #400
- Mask overall CI status on main branch
- Badges!!!

Depends on: #361 #358

LecrisUT committed:
Separate pre-commit step

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
fix concurrency references

Concurrency should be checked only on the top-level CI/caller

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Simplify CI names

Make the CI report more navigable

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Separate coverage step

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Mask experimental runs

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Badges!

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT commented on README.md:
```diff
@@ -1,9 +1,21 @@
 
+[![CI Status][ci-badge]][ci-link]
+[![Documentation Status][rtd-badge]][rtd-link]
+[![Codecov Status][codecov-badge]][codecov-link]
```
Codecov badge needs a token

LecrisUT commented on README.md:
```diff
@@ -1,9 +1,21 @@
+
+![License Status][license-badge]
+![CMake Status][cmake-badge]
+[![Python Versions][python-version]][pypi-link]
```
Python version badge will be fixed automatically in next pypi release

LecrisUT commented on README.md:
```diff
@@ -1,9 +1,21 @@
-
 # Spglib
 
+[![CI Status][ci-badge]][ci-link]
+[![Documentation Status][rtd-badge]][rtd-link]
+[![Codecov Status][codecov-badge]][codecov-link]
+
+![License Status][license-badge]
+![CMake Status][cmake-badge]
+[![Python Versions][python-version]][pypi-link]
+
+[![Release Status][release-badge]][release-link]
+[![Conda Status][conda-badge]][conda-link]
+[![PyPI Status][pypi-badge]][pypi-link]
+[![Fedora Status][fedora-badge]][fedora-link]
+
+[![PyPI Downloads][pypi-downloads]][pypi-link]
+[![Conda Downloads][conda-downloads]][conda-link]
```
Most badges are from https://shields.io/badges. There could be more interesting ones to find. I think a social one would be good, but there is no matrix/discord room etc. Doesn't have to be spglib specific, just a place where developers can be contacted in real-time

lan496 replied:
I think the GitHub issue and discussion work as such a community room now. So, I do not feel the need for another social one so much.

lan496 commented:
@LecrisUT I've received "Request to install Codecov in spglib". Is it needed for the codecov badge? I remember this project is already integrated with codecov. What is the difference?

LecrisUT commented:
> @LecrisUT I've received "Request to install Codecov in spglib". Is it needed for the codecov badge? I remember this project is already integrated with codecov. What is the difference?

Can you get the badge api-token without installing codecov app? That's the main thing I'm trying with that.

lan496 commented:
Do you need the badge api token as GitHub's Action secrets or raw string?

LecrisUT commented:
> Do you need the badge api token as GitHub's Action secrets or raw string?

Raw string, it's meant to be public and used in the README.md

lan496 commented:
I see. Then, here it is.
```
[![codecov](https://codecov.io/gh/spglib/spglib/graph/badge.svg?token=G91xAWu1JT)](https://codecov.io/gh/spglib/spglib)
```

LecrisUT committed:
Change step_test matrix

- Reduce python tests: Only minimal + latest + future python version
- Fix os toolchains

Signed-off-by: Cristian Le <git@lecris.dev>

LecrisUT committed:
Disable `windows-2019` and python --pre to be investigated later

Signed-off-by: Cristian Le <git@lecris.dev>

LecrisUT commented:
All badges except for Python version are working. Just 2 things for this PR:
- Feedback on namings:
  - I am considering `CI / 📘 docs / Sphinx-html` instead, but would that be confusing?
  - Would `🧪` be sufficient to indicate experimental jobs?
  - `+` sign is excessive?
  - Other suggestions for emojis?
- When ready to merge, the `pass` job changed name a bit.

lan496 commented:
> Mask overall CI status on main branch

Can you explain this a little more? Are these badges reflected only on main branch?

lan496 commented on README.md:
```diff
@@ -177,3 +189,23 @@ improvements. See [Contributing.md](Contributing.md) for more details.
 documentation. See the [doc documentation](doc/README.md) for more information on
 how to contribute to the documentation.
+
+[ci-badge]: https://github.com/LecrisUT/CMake-Template/actions/workflows/ci.yaml/badge.svg?branch=main&event=push
```
Is this link correct? It seems to link to your action.

LecrisUT replied:
Oops, thanks for the catch :D

lan496 commented:
> Feedback on namings:

I think the current emojis are nice!

LecrisUT commented:
> > Mask overall CI status on main branch
> 
> Can you explain this a little more? Are these badges reflected only on main branch?

It's easiest with an example:
If you open the CI checks for [this PR](https://github.com/LecrisUT/CMake-Template/pull/10), you will see `windows` is failing, but the [merge commit](https://github.com/LecrisUT/CMake-Template/commit/a4f73dbd53d8cb98e6325296c2905b2ed40dcaa3) is showing it as passing. That is because I am masking the true result there so that on the [main page](https://github.com/LecrisUT/CMake-Template) the first commit is shown with a check mark.

lan496 commented on .github/workflows/step_test.yaml:
```diff
@@ -1,52 +1,50 @@
-        toolchain: [ gcc, llvm, intel, windows, macos ]
-        python-version: [ "3.8", "3.9", "3.10", "3.11", "3.12"]
+        toolchain: [ gcc, llvm, intel ]
+        python-version: [ "3.8", "3.x" ]
```
Note: "3.x" specifies the latest stable version.
https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#using-the-python-version-input

lan496 commented:
I see. Thanks, I've left a few comments.

lan496 commented:
> Add a static-analysis checker. I was experimenting with sonarcloud at CMake-Template repo. I think it could be useful to include it as an intermediate to clang-tidy. It is also polyglot, so it might pick up on more nieche problems

Is it included?

LecrisUT commented:
> > Add a static-analysis checker. I was experimenting with sonarcloud at CMake-Template repo. I think it could be useful to include it as an intermediate to clang-tidy. It is also polyglot, so it might pick up on more nieche problems
> 
> Is it included?

Not yet, I will postpone it for after release #400

lan496 commented on README.md:
```diff
@@ -177,3 +189,23 @@ improvements. See [Contributing.md](Contributing.md) for more details.
+
+[ci-badge]: https://github.com/spglib/spglib/actions/workflows/ci.yaml/badge.svg?branch=develop&event=push
+[ci-link]: https://github.com/spglib/spglib/actions/workflows/ci.yaml?query=branch%3Adevelop+event%3Apush
+[cmake-badge]: https://img.shields.io/badge/CMake-3.15-blue?logo=data:image/svg%2bxml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGRhdGEtbmFtZT0iTGF5ZXIgMiIgdmlld0JveD0iMCAwIDU0NS41IDU0NS45Ij48cGF0aCBkPSJNNTQ2IDUzNCAyODIgOWwzOSA0MzQgMjI1IDkxeiIgZGF0YS1uYW1lPSJyZWQgcmlnaHQiIHN0eWxlPSJmaWxsOiNmMzI3MzU7c3Ryb2tlLXdpZHRoOjAiLz48cGF0aCBkPSJNNTI1IDU0NiAxNzAgNDAzIDEgNTQ2aDUyNHoiIGRhdGEtbm ...
```
Note: We have to manually update the badge for the CMake minimum version.

LecrisUT replied:
Yes, specifically only the `3.15` text and only when we bump minimum version

lan496 commented:
@LecrisUT Can you address the branch conflicts?

LecrisUT commented:
> @LecrisUT Can you address the branch conflicts?

@lan496 Done. I forgot, you need to switch the required CI for the auto-merge

lan496 commented:
Current branch rules
![image](https://github.com/spglib/spglib/assets/7272505/1f594d3c-b3e1-4632-9474-dd110f77f18f)
